src/utilities/createRouter.ts

```diff
@@ -5,58 +5,76 @@ import { Routes, Router, RouterOptions, RouterImplementation } from '@/types'
 import { createRouterPush } from '@/utilities/createRouterPush'
 import { createRouterReject } from '@/utilities/createRouterReject'
 import { createRouterReplace } from '@/utilities/createRouterReplace'
 import { createRouterResolve } from '@/utilities/createRouterResolve'
 import { createRouterRoute } from '@/utilities/createRouterRoute'
 import { getInitialUrl } from '@/utilities/getInitialUrl'
-import { executeMiddleware } from '@/utilities/middleware'
-import { getRouterRouteForUrl } from '@/utilities/routes'
+import { OnMiddlewareError, executeMiddleware } from '@/utilities/middleware'
+import { getRouteHooks, getRouterRouteForUrl } from '@/utilities/routes'
 
 export function createRouter<const T extends Routes>(routes: T, options: RouterOptions = {}): Router<T> {
   const resolved = resolveRoutes(routes)
   const resolve = createRouterResolve({ resolved })
+  const {
+    onBeforeRouteEnter,
+    onBeforeRouteLeave,
+    onBeforeRouteUpdate,
+    addRouteHook,
+    hooks,
+  } = createRouterHooks()
+
+  const onMiddlewareError: OnMiddlewareError = (error) => {
+    if (error instanceof RouterRejectionError) {
+      reject(error.type)
+    }
+
+    if (error instanceof RouterPushError) {
+      const [source, options] = error.to
+      const url = resolve(source)
+
+      navigation.update(url, options)
+      return
+    }
+
+    if (error instanceof RouterReplaceError) {
+      const [source, options] = error.to
+      const url = resolve(source, options)
+
+      navigation.update(url, { replace: true })
+      return
+    }
+
+    if (!(error instanceof RouterRejectionError)) {
+      throw error
+    }
+  }
 
   const onLocationUpdate = async (url: string): Promise<void> => {
-    const matched = getRouterRouteForUrl(resolved, url)
+    const to = getRouterRouteForUrl(resolved, url)
+    const from = isRejectionRoute(route) ? null : route
 
-    if (!matched) {
+    if (!to) {
       return reject('NotFound')
     }
 
-    try {
-      await executeMiddleware({
-        to: matched,
-        from: isRejectionRoute(route) ? null : route,
-      })
-    } catch (error) {
-      if (error instanceof RouterRejectionError) {
-        reject(error.type)
-      }
-
-      if (error instanceof RouterPushError) {
-        const [source, options] = error.to
-        const url = resolve(source, options)
-
-        navigation.update(url, options)
-        return
-      }
-
-      if (error instanceof RouterReplaceError) {
-        const [source, options] = error.to
-        const url = resolve(source, options)
-
-        navigation.update(url, { replace: true })
-        return
-      }
-
-      if (!(error instanceof RouterRejectionError)) {
-        throw error
-      }
+    const success = await executeMiddleware({
+      middleware: [
+        ...hooks.before,
+        ...getRouteHooks(to, 'before'),
+      ],
+      to,
+      from,
+      onMiddlewareError,
+    })
+
+    if (!success) {
+      return
```

Comment:
executeMiddleware use to just accept `to`, and `from` and then it would grab the middleware from the `to` route and throw errors. 

Now you have to tell it what middleware to execute AND how to handle errors. I was looking for a way to abstract the error handling because I'll need to execute middleware in two places (for before and after hooks) but use the same error handling logic. 

I think passing in an error handler makes sense because its expected that this function will throw errors. So it should never be used without an error handler.